Keycloak benchmark dataset provider unable to create users

### Describe the bug

I am able to use Keycloak benchmark dataset provider 26.0 and 26.2 to create realm and clients but when i try to create users, it gives me error.



### Version

26.0.0

### Expected behavior

Create users successfully

### Actual behavior

2025-11-07 16:02:24,196 INFO  [org.keycloak.benchmark.dataset.DatasetResourceProvider] (Thread-23) Cached realm realm-0, Time since start of the task 'Creation of 5 users in the realm realm-0': 77 ms
2025-11-07 16:02:24,197 INFO  [org.keycloak.benchmark.dataset.ExecutorHelper] (Thread-23) Waiting for tasks to complete successfully
2025-11-07 16:02:24,314 ERROR [org.keycloak.benchmark.dataset.DatasetResourceProvider] (Thread-23) unable to complete task: java.lang.RuntimeException: Some futures failed
        at org.keycloak.benchmark.dataset.ExecutorHelper.waitForAllToFinish(ExecutorHelper.java:75)
        at org.keycloak.benchmark.dataset.DatasetResourceProvider.createUsersImpl(DatasetResourceProvider.java:424)
        at org.keycloak.benchmark.dataset.DatasetResourceProvider.lambda$createUsers$11(DatasetResourceProvider.java:393)
        at java.base/java.lang.Thread.run(Thread.java:840)
        Suppressed: java.util.concurrent.ExecutionException: java.lang.ArithmeticException: / by zero
                at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
                at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
                at org.keycloak.benchmark.dataset.ExecutorHelper.waitForAllToFinish(ExecutorHelper.java:69)
                ... 3 more
        Caused by: java.lang.ArithmeticException: / by zero
                at org.keycloak.benchmark.dataset.DatasetResourceProvider.createUsers(DatasetResourceProvider.java:1169)
                at org.keycloak.benchmark.dataset.DatasetResourceProvider.lambda$addUserCreationTasks$12(DatasetResourceProvider.java:448)
                at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:263)
                at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:392)
                at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:262)
                at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithTimeout(KeycloakModelUtils.java:413)
                at org.keycloak.benchmark.dataset.ExecutorHelper.lambda$addTaskRunningInTransaction$0(ExecutorHelper.java:58)
                at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
                at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
                at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
                at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
                ... 1 more


### How to Reproduce?

Install keycloak benchmark dataset provider 
https://www.keycloak.org/keycloak-benchmark/dataset-guide/latest/using-provider  

https://<keycloak_address>/realms/master/dataset/create-users?count=5&realm-name=realm-0

### Anything else?

Hope someone can share some insights on this. Thank you in advance!

--- a/dataset/src/main/java/org/keycloak/benchmark/dataset/DatasetResourceProvider.java
+++ b/dataset/src/main/java/org/keycloak/benchmark/dataset/DatasetResourceProvider.java
@@ -63,6 +63,7 @@ import org.keycloak.services.managers.UserSessionManager;
 import org.keycloak.services.resource.RealmResourceProvider;
 
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
@@ -1134,6 +1135,10 @@ public class DatasetResourceProvider implements RealmResourceProvider {
         RealmModel realm = session.realms().getRealm(context.getRealm().getId());
         DatasetConfig config = context.getConfig();
 
+        warnMissingResource(realm.getName(), "realm-roles", context.getRealmRoles(), config.getRealmRolesPerUser());
+        warnMissingResource(realm.getName(), "client-roles", context.getClientRoles(), config.getClientRolesPerUser());
+        warnMissingResource(realm.getName(), "groups", context.getGroups(), config.getGroupsPerUser());
+
         for (int i = startIndex; i < endIndex; i++) {
             String username = config.getUserPrefix() + i;
             UserModel user = session.users().addUser(realm, username);
@@ -1154,9 +1159,6 @@ public class DatasetResourceProvider implements RealmResourceProvider {
 
                     logger.tracef("Assigned role %s to the user %s", context.getRealmRoles().get(roleIndex).getName(), user.getUsername());
                 }
-            } else if (config.getRealmRolesPerUser() > 0) {
-                logger.warnf("realm-roles-per-user=%d but no realm roles defined in realm %s, skipping role assignment for user %s",
-                      config.getRealmRolesPerUser(), realm.getName(), user.getUsername());
             }
 
             // Assign a client role to a user if any exist in the realm
@@ -1168,9 +1170,6 @@ public class DatasetResourceProvider implements RealmResourceProvider {
 
                     logger.tracef("Assigned role %s to the user %s", context.getClientRoles().get(roleIndex).getName(), user.getUsername());
                 }
-            } else if (config.getClientRolesPerUser() > 0) {
-                logger.warnf("client-roles-per-user=%d but no client roles defined in realm %s, skipping client role assignment for user %s",
-                      config.getClientRolesPerUser(), realm.getName(), user.getUsername());
             }
 
             // Assign a group to a user if any exist
@@ -1183,15 +1182,18 @@ public class DatasetResourceProvider implements RealmResourceProvider {
 
                     logger.tracef("Assigned group %s to the user %s", context.getGroups().get(groupIndex).getName(), user.getUsername());
                 }
-            } else if (config.getGroupsPerUser() > 0) {
-                logger.warnf("groups-per-user=%d but no groups defined in realm %s, skipping joining group for user %s",
-                      config.getGroupsPerUser(), realm.getName(), user.getUsername());
             }
 
             context.incUserCount();
         }
     }
 
+    private void warnMissingResource(String realm, String resourceName, Collection<?> resource, Integer expectedCount) {
+        if (resource.isEmpty() && expectedCount > 0) {
+            logger.warnf("%1$s-per-user=%2$d but no %1$s are defined in realm %3$s, skipping %1$s assignment for all users",
+                  resourceName, expectedCount, realm);
+        }
+    }
 
     private void cacheRealmAndPopulateContext(RealmContext context) {
         DatasetConfig config = context.getConfig();

